src/data_info/services/path_lists.clj

```diff
@@ -0,0 +1,124 @@
+   :type       (case type "dataobject" :file "collection" :dir)
+   :path       full_path
+   :label      base_name
+   :infoType   info_type
+   :permission (perms/fmt-perm access_type_id)})
+
+(defn- filter-entity-type
+  [recursive? folders-only?]
+  (cond
+    (and recursive? (not folders-only?)) :any
+    folders-only?                        :folder
+    :else                                :file))
+
+(defn- folder-listing
+  "Fetches a folder's contents, listing files, folders, or both depending on the given parameters."
+  [user info-types folders-only? recursive? path]
+  (map fmt-entry
+       (icat/paged-folder-listing
+         :user           user
+         :zone           (cfg/irods-zone)
+         :folder-path    path
+         :info-types     info-types
+         :entity-type    (filter-entity-type recursive? folders-only?)
+         :sort-column    :full-path
+         :sort-direction :asc
+         :limit          nil
+         :offset         0)))
+
+(defn- list-item-with-subitems
+  [user info-types folders-only? recursive? {:keys [path] :as data-item}]
+  "If given a file, returns that file as the only item in a list.
+   If given a folder, returns that folder in a list, and the list may include the folder's subfolder/files
+   depending on the given parameters."
+  (let [sub-listing (when (and recursive? (stat-is-dir? data-item))
+                      (mapcat (partial list-item-with-subitems user info-types folders-only? recursive?)
+                              (folder-listing user info-types folders-only? recursive? path)))]
+    (concat [data-item] sub-listing)))
+
+(defn- keep-top-level-file?
+  "A filter predicate to keep only files with an info-type that matches one in the given `info-types` list,
+   or to keep all files if `info-types` is empty."
+  [info-types]
+  (if (empty? info-types)
+    (constantly true)
+    (comp (partial some (set info-types)) list :infoType)))
+
+(defn- label-matches?
+  "Returns true if the given `name-pattern` matches the given label, or if the given `name-pattern` is blank."
+  [name-pattern label]
+  (or (string/blank? name-pattern)
+      (re-find (re-pattern name-pattern) label)))
+
+(defn- keep-data-item?
+  "Returns true if the given `data-item` has permissions set,
+   its label passes the `label-matches?` check,
+   and if it's a file or folder depending on the given `folders-only?` flag."
+  [name-pattern folders-only? {:keys [label permission] :as data-item}]
+  (and permission
+       (label-matches? name-pattern label)
+       (if (stat-is-dir? data-item)
+         folders-only?
+         (not folders-only?))))
+
+(defn- paths->ht-path-list
+  "Filters the given paths and returns a string of these paths appended to an HT Path List header.
+   Throws an error if the filtering params result in no matching paths."
+  [cm user name-pattern info-types folders-only? recursive? paths]
+  (let [{folder-paths true file-paths false} (group-by (partial item-info/is-dir? cm) paths)
```

Comment:
This `item-info/is-dir?` will possibly slow things down for you. With this, you're basically already doing a decent chunk of a `stat` (the same call that `is-dir?` uses provides path, type, created/modified dates, file size, and checksum). I'm not sure if there's a sensible way to rewrite this to just fetch stats in the first place to reduce processing further down, though. The section below wants infoType, which would only apply to files so would be fine to add, but it also wants permission which is an extra iRODS call that only the files section below looks like it uses. Anyway, food for thought maybe?